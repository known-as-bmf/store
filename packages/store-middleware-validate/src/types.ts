export type Validator<S> = (state: S) => boolean;
