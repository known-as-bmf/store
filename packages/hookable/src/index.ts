export { createHookable } from './hookable';
export { every, once } from './hook-subscribers';
export {
  Hook,
  HookSubscriber,
  Hooks,
  TapInput,
  TapOutput,
  TransformInput,
  TransformOutput,
} from './types';
