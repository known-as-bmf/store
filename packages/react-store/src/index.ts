export { useStore } from './useStore';
export { useNewStore } from './useNewStore';
