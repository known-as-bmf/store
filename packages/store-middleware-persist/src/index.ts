export { PersistOptions } from './types';
export { persistMiddleware } from './persist';
