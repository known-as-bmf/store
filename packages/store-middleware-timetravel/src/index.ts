export { TimetravelOptions } from './types';
export { timetravelMiddleware, undo, redo, derefHistory } from './timetravel';
