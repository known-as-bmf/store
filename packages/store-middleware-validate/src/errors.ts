import { Validator } from './types';

export const errors = {
  validationFailed: <S>(state: S, validator: Validator<S>): string =>
    `Validation failed for state:\n\n${JSON.stringify(
      state,
      undefined,
      ' '
    )}\n\nFailed validator:\n\n${validator.toString()}`,
};
