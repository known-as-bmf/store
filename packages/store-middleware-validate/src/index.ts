export { Validator } from './types';
export { validatorMiddleware } from './validator';
