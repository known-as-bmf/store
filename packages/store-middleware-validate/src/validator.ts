import { Middleware } from '@known-as-bmf/store';

import { Validator } from './types';
import { errors } from './errors';

/**
 * Create a new validation middleware.
 * @param validators A list of validators that will be checked against every
 * time the state is being changed.
 */
export const validatorMiddleware = <S>(
  validators: Validator<S>[]
): Middleware<S> => (_, hooks) => {
  hooks.stateWillChange(() => (futureState) => {
    const failedValidator = validators.find(
      (validator) => !validator(futureState)
    );

    if (failedValidator) {
      throw new Error(errors.validationFailed(futureState, failedValidator));
    }
  });
};
