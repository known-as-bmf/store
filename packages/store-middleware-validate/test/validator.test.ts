import { of, set, swap, deref } from '@known-as-bmf/store';

import { validatorMiddleware } from '../src';

describe('validatorMiddleware', () => {
  it('should be a function', () => {
    expect(validatorMiddleware).toBeDefined();
    expect(validatorMiddleware).toBeInstanceOf(Function);
  });
});

describe('of', () => {
  it('should throw error if initial state is invalid', () => {
    expect(() =>
      of({ test: 10 }, validatorMiddleware([(s) => s.test < 5]))
    ).toThrowErrorMatchingSnapshot();
  });
});

describe('swap', () => {
  it('should throw error if new state is invalid', () => {
    const store = of({ test: 0 }, validatorMiddleware([(s) => s.test < 5]));

    expect(() =>
      swap(store, (s) => ({ ...s, test: 10 }))
    ).toThrowErrorMatchingSnapshot();

    expect(deref(store)).toEqual({ test: 0 });
  });
});

describe('set', () => {
  it('should throw error if new state is invalid', () => {
    const store = of({ test: 0 }, validatorMiddleware([(s) => s.test < 5]));

    expect(() => set(store, { test: 10 })).toThrowErrorMatchingSnapshot();

    expect(deref(store)).toEqual({ test: 0 });
  });
});
