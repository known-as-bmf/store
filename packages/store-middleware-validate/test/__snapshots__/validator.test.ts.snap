// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`of should throw error if initial state is invalid 1`] = `
"Validation failed for state:

{
 \\"test\\": 10
}

Failed validator:

function (s) { return s.test < 5; }"
`;

exports[`set should throw error if new state is invalid 1`] = `
"Validation failed for state:

{
 \\"test\\": 10
}

Failed validator:

function (s) { return s.test < 5; }"
`;

exports[`swap should throw error if new state is invalid 1`] = `
"Validation failed for state:

{
 \\"test\\": 10
}

Failed validator:

function (s) { return s.test < 5; }"
`;
